File: scripts/iroh-qualification/prototype.ts
import assert from "node:assert/strict";
import { generateKeyPairSync, randomUUID } from "node:crypto";
import { mkdirSync, mkdtempSync, writeFileSync } from "node:fs";
import { resolve, join } from "node:path";
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";
import { z } from "zod";
import { RelayClient } from "../../bridge/src/relay-client";
import { MessageBus } from "../../bridge/src/message-bus";
import { generateEphemeralKeypair } from "../../bridge/src/key-exchange";
import { TrustedPeersProvider } from "../../bridge/src/trusted-peers";
import { createMessage, AbMessageSchema } from "../../bridge/src/protocol";
import { TerminalSession } from "../../bridge/src/terminal-session";
import { TerminalFrameSource } from "../../bridge/src/terminal-frames/source";
import { TERMINAL_PROTOCOL_VERSION } from "../../bridge/src/terminal-frames/protocol";
import { setLogLevel } from "../../bridge/src/logger";
import { ClientMessage, RouteHeader, FrameKind, decodeRouteFrame, encodeRouteFrame } from "../../packages/antgrid-wire/src/index";
import { Records } from "./records";

const { Endpoint, presetMinimal } = require("@number0/iroh/index.js");
const machineId = "iroh-prototype-machine";
const appId = "iroh-prototype-app";
const alpn = Array.from(Buffer.from("antgrid/peer/1"));
const readySchema = z.object({
  endpointId: z.string().regex(/^[0-9a-f]{64}$/),
  publicKey: z.string().base64(),
});
const resultSchema = z.object({
  check: z.literal("encrypted-terminal-pass"),
  establishments: z.literal(2), projects: z.literal(2),
  sameConnection: z.literal(true),
});

setLogLevel("fatal");
const executable = process.env.IROH_QUALIFICATION_DART;
if (!executable) throw new Error("Set IROH_QUALIFICATION_DART to dart.exe or the compiled prototype client");
const compiled = process.env.IROH_QUALIFICATION_COMPILED === "1";
const testCase = z.enum(["terminal", "bad-agent-key", "bad-app-key", "oversize-record", "wrong-destination", "extra-stream", "disconnect",
  "client-oversize-record", "client-wrong-destination", "client-extra-stream"])
  .parse(process.env.IROH_QUALIFICATION_CASE ?? "terminal");
const base = compiled ? resolve(".tmp/iroh-qualification") : resolve(import.meta.dir, "../../.tmp/iroh-qualification");
mkdirSync(base, { recursive: true });
const scratch = mkdtempSync(join(base, "prototype-"));
const key = generateKeyPairSync("ed25519");
const machinePublic = key.publicKey.export({ type: "spki", format: "der" }).subarray(-32).toString("base64");
const machineSeed = key.privateKey.export({ type: "pkcs8", format: "der" }).subarray(-32).toString("base64");
const child = spawn(executable, compiled ? [] : ["run", "bin/prototype.dart"], {
  cwd: compiled ? undefined : resolve(import.meta.dir, "dart"),
  stdio: ["pipe", "pipe", "inherit"],
});
const exited = new Promise<number | null>((resolve, reject) => {
  child.once("exit", resolve); child.once("error", reject);
});
void exited.catch(() => {});
const failed = Promise.withResolvers<never>();
void failed.promise.catch(() => {});
const fail = (error: unknown) => failed.reject(error);
const watchdog = setTimeout(() => {
  for (const project of projects.values()) project.terminal.kill();
  child.kill();
  console.error("FAIL: qualification prototype exceeded 60 seconds");
  process.exit(1);
}, 60_000);
const projects = new Map<string, {
  streamId: string; terminal: TerminalSession; source: TerminalFrameSource;
  bus: MessageBus; timer: ReturnType<typeof setInterval>;
}>();
let endpoint: any;
let connection: any;
let records: Records | undefined;
let client: RelayClient | undefined;
let server: ReturnType<typeof Bun.serve> | undefined;
let fixtureSocket: any;
let establishments = 0;
let terminalInputs = 0;
let shuttingDown = false;

try {
  const lines = createInterface({ input: child.stdout });
  const iterator = lines[Symbol.asyncIterator]();
  const first = await Promise.race([iterator.next(), failed.promise]);
  assert.equal(first.done, false);
  const app = readySchema.parse(JSON.parse(first.value!));
  console.error("prototype: Dart identity ready");
  const inventoryFile = join(scratch, "peers.json");
  writeFileSync(inventoryFile, JSON.stringify([{ deviceId: appId,
    ed25519Pub: testCase === "bad-app-key" ? machinePublic : app.publicKey }]));
  const trustedPeers = new TrustedPeersProvider({ filePath: inventoryFile,
    licenseApiUrl: "http://127.0.0.1:1", getToken: () => "qualification-fixture" });
  const control = new MessageBus();
  const pathToken = randomUUID();
  const authenticated = Promise.withResolvers<void>();

  // Local socket glue preserves the production RelayClient driver. It never
  // reaches LocalListener/HostServer's loopback command-authorization exemption.
  server = Bun.serve({
    hostname: "127.0.0.1", port: 0,
    fetch(request, server) {
      if (new URL(request.url).pathname !== `/${pathToken}` || fixtureSocket) return new Response(null, { status: 403 });
      return server.upgrade(request, { data: null }) ? undefined : new Response(null, { status: 400 });
    },
    websocket: {
      open(ws) { fixtureSocket = ws; },
      message(ws, raw) {
        try {
          if (typeof raw === "string") {
            const message = ClientMessage.parse(JSON.parse(raw));
            if (message.type === "hello") {
              assert.equal(message.deviceId, machineId);
              assert.equal(message.publicKey, machinePublic);
              ws.send(JSON.stringify({ type: "welcome", deviceId: machineId,
                epoch: message.epoch, serverTime: new Date().toISOString() }));
            } else if (message.type === "stream-open") {
              ws.send(JSON.stringify({ type: "stream-opened", streamId: message.streamId }));
            } else if (message.type === "ping") ws.send(JSON.stringify({ type: "pong" }));
            return;
          }
          const decoded = decodeRouteFrame(raw);
          assert.equal(RouteHeader.parse(decoded.header).to, appId);
          assert.ok(records, "Payload cannot precede the selected Iroh connection");
          records.send(raw);
        } catch (error) { fail(error); }
      },
      close() { if (!shuttingDown) fail(new Error("Fixture control socket lost")); },
    },
  });
  client = new RelayClient({
    url: `ws://127.0.0.1:${server.port}/${pathToken}`,
    identity: { deviceId: machineId, deviceName: "qualification", createdAt: "",
      ed25519PublicKey: machinePublic, ed25519PrivateKey: machineSeed },
    generateKeypair: generateEphemeralKeypair, trustedPeers,
    autoReconnect: false, getLicenseToken: () => "qualification-fixture",
    onAuthenticated: () => authenticated.resolve(),
    onHandshakeComplete: () => { establishments++; },
    onError: (code) => fail(new Error(code)),
  });
  client.setBus(control);
  const snapshot = (bus: MessageBus, msg: any) => {
    if (msg.type !== "request") return false;
    if (msg.method === "qualification.drop") {
      connection.close(42n, []);
      return true;
    }
    bus.publish(createMessage("response", { requestId: msg.requestId, ok: true, result: { frames: [] } }), "control");
    return true;
  };
  control.setInboundHandler((message, _channel, source) => {
    try {
      assert.equal(source, "relay");
      if (snapshot(control, message)) return;
      if (message.type !== "project:start") return;
      const projectId = message.projectId;
      assert.ok(["prototype-a", "prototype-b"].includes(projectId));
      let project = projects.get(projectId);
      if (!project) {
        const bus = new MessageBus();
        const screen = new TerminalFrameSource(100, 20);
        const runId = randomUUID();
        let attachmentId = randomUUID();
        let sequence = 0;
        let subscribed = false;
        let acknowledged = -1;
        const terminal = new TerminalSession({
          terminalId: projectId, cols: 100, rows: 20, cwd: scratch,
          command: process.platform === "win32" ? process.env.ComSpec ?? "cmd.exe" : "/bin/sh",
          args: process.platform === "win32" ? ["/d", "/q"] : [],
          onMessage(message) { if (message.type === "terminal:output") screen.feed(message.data); },
        });
        const handle = client!.attachStream(bus, { projectId,
          mayDeliver: () => !shuttingDown,
          mayAcceptFrom: (peer) => peer?.checkoutRouting ? null : { code: "NOT_ALLOWED", message: "checkout capability required" },
        });
        bus.setInboundHandler((message, _channel, source) => {
          try {
            assert.equal(source, "relay");
            if (snapshot(bus, message)) return;
            const msg = AbMessageSchema.parse(message);
            if (msg.type === "terminal:subscribe") {
              assert.equal(msg.terminalId, projectId);
              assert.equal(msg.version, TERMINAL_PROTOCOL_VERSION);
              attachmentId = randomUUID(); sequence = 0; acknowledged = -1;
              subscribed = true;
              bus.publish(createMessage("terminal:subscribed", { terminalId: projectId, runId, attachmentId,
                requestId: msg.requestId, version: TERMINAL_PROTOCOL_VERSION }), "control");
            } else if (msg.type === "terminal:ack" && msg.attachmentId === attachmentId) acknowledged = msg.sequence;
            else if (msg.type === "terminal:input") {
              assert.equal(msg.terminalId, projectId);
              const allowed = process.platform === "win32"
                ? ["set /a 731*37\r", "set /a 947*19\r"]
                : ["echo $((731*37))\n", "echo $((947*19))\n"];
              assert.ok(allowed.includes(msg.data));
              terminalInputs++;
              terminal.write(msg.data);
            }
          } catch (error) { fail(error); }
        });
        const timer = setInterval(() => {
          try {
            if (!subscribed || sequence - acknowledged > 2) return;
            const frame = screen.capture(performance.now());
            if (!frame) return;
            bus.publish(AbMessageSchema.parse(createMessage("terminal:frame", {
              ...frame, terminalId: projectId, runId, attachmentId, sequence: sequence++,
            })), "control");
          } catch (error) { fail(error); }
        }, 30);
        project = { streamId: handle.streamId, terminal, source: screen, bus, timer };
        projects.set(projectId, project);
        terminal.spawn();
      }
      control.publish(createMessage("stream-ready", { projectId, streamId: project.streamId }), "control");
    } catch (error) { fail(error); }
  });
  client.connect();
  await Promise.race([authenticated.promise, failed.promise]);
  console.error("prototype: bridge driver ready");
  const builder = Endpoint.builder();
  presetMinimal(builder);
  builder.bindAddr("127.0.0.1:0"); builder.alpns([alpn]);
  endpoint = await builder.bind();
  child.stdin.end(JSON.stringify({ endpointId: endpoint.id().toString(),
    addresses: endpoint.boundSockets().filter((x: string) => x.startsWith("127.0.0.1:")),
    machinePublic: testCase === "bad-agent-key" ? app.publicKey : machinePublic,
    machineId, appId, windows: process.platform === "win32", testCase,
    terminalVersion: TERMINAL_PROTOCOL_VERSION }) + "\n");
  const incoming = await endpoint.acceptNext();
  assert.ok(incoming);
  connection = await (await incoming.accept()).connect();
  console.error("prototype: Iroh connected");
  assert.equal(connection.remoteId().toString(), app.endpointId);
  assert.deepEqual(connection.alpn(), alpn);
  const stream = await connection.acceptBi();
  records = new Records(stream, fail);
  void connection.acceptBi().then(() => fail(new Error("EXTRA_STREAM")), () => {});
  void connection.acceptUni().then(() => fail(new Error("EXTRA_STREAM")), () => {});
  if (testCase.startsWith("client-")) {
    if (testCase === "client-oversize-record") await stream.send.writeAll([255, 255, 255, 255]);
    else if (testCase === "client-wrong-destination") {
      const raw = encodeRouteFrame({ type: "message", to: "wrong-app", channel: "control" }, Buffer.from("{}"), FrameKind.handshake);
      const prefix = Buffer.alloc(4); prefix.writeUInt32BE(raw.length);
      await stream.send.writeAll([...prefix, ...raw]);
    } else {
      const extra = await connection.openBi(); await extra.send.writeAll([1]);
    }
  } else void (async () => {
    try {
      while (!shuttingDown) {
        const frame = decodeRouteFrame(await records!.read());
        const header = RouteHeader.parse(frame.header);
        if (header.to !== machineId) throw new Error("INVALID_DESTINATION");
        fixtureSocket.send(encodeRouteFrame({ type: "message", from: appId,
          channel: header.channel, ts: Date.now() }, frame.payload, frame.kind));
      }
    } catch (error) {
      const expectedClose = String(error).includes("ApplicationClosed") ||
        (testCase === "disconnect" && String(error).includes("LocallyClosed"));
      if (!shuttingDown && !expectedClose) fail(error);
    }
  })();
  const result = await Promise.race([iterator.next(), failed.promise]);
  assert.equal(result.done, false);
  if (testCase === "terminal") {
    resultSchema.parse(JSON.parse(result.value!));
    assert.equal(establishments, 2);
    assert.equal(terminalInputs, 3, "No duplicate terminal inputs across project switches or rekey");
    assert.equal(projects.size, 2);
  } else if (testCase.startsWith("client-")) {
    z.object({ check: z.literal("peer-protocol-rejected"), testCase: z.literal(testCase) }).parse(JSON.parse(result.value!));
    assert.equal(establishments, 0); assert.equal(terminalInputs, 0); assert.equal(projects.size, 0);
  } else if (testCase === "disconnect") {
    z.object({ check: z.literal("pending-failed"), code: z.literal("E_SESSION_DOWN") }).parse(JSON.parse(result.value!));
    assert.equal(establishments, 1); assert.equal(terminalInputs, 0);
  } else {
    z.object({ check: z.literal("identity-rejected"), testCase: z.literal(testCase) }).parse(JSON.parse(result.value!));
    assert.equal(establishments, 0); assert.equal(terminalInputs, 0); assert.equal(projects.size, 0);
  }
  shuttingDown = true;
  lines.close();
  assert.equal(await exited, 0);
  console.log(JSON.stringify({ check: "encrypted-terminal-prototype", testCase, status: "pass", establishments,
    projects: projects.size, terminalInputs, framesSent: records.sent, framesReceived: records.received,
    peakQueuedBytes: records.peakQueuedBytes, profile: "loopback-no-relay",
    fixtures: ["account inventory", "control admission", "project catalog and dispatch"] }));
} catch (error) {
  const expected: Record<string, string> = {
    "oversize-record": "INVALID_RECORD_LENGTH", "wrong-destination": "INVALID_DESTINATION", "extra-stream": "EXTRA_STREAM",
  };
  if (!(error instanceof Error) || error.message !== expected[testCase]) throw error;
  assert.equal(terminalInputs, 0); assert.equal(projects.size, 0);
  console.log(JSON.stringify({ check: "native-protocol-rejection", testCase, status: "pass", reason: error.message, terminalInputs }));
} finally {
  shuttingDown = true;
  records?.close(); connection?.close(0n, []); child.kill();
  for (const project of projects.values()) {
    clearInterval(project.timer); project.source.dispose(); await project.terminal.close(0);
  }
  client?.close(); server?.stop(true); await endpoint?.close(); clearTimeout(watchdog);
}

File: scripts/iroh-qualification/dart/bin/prototype.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';
import 'package:antgrid_relay_client/antgrid_relay_client.dart';
import 'package:iroh_quic/iroh_quic.dart' as iroh;
import 'package:iroh_quic/src/rust/frb_generated.dart' show RustLib;
import 'package:antgrid_iroh_qualification/peer_link.dart';

String uuid() {
  final r = Random.secure();
  final b = List.generate(16, (_) => r.nextInt(256));
  b[6] = (b[6] & 15) | 64;
  b[8] = (b[8] & 63) | 128;
  final h = b.map((n) => n.toRadixString(16).padLeft(2, '0')).join();
  return '${h.substring(0, 8)}-${h.substring(8, 12)}-${h.substring(12, 16)}-${h.substring(16, 20)}-${h.substring(20)}';
}

Map<String, dynamic> message(String type, Map<String, dynamic> fields) => {
  'type': type,
  'id': uuid(),
  'timestamp': DateTime.now().millisecondsSinceEpoch,
  ...fields,
};

Future<void> main() async {
  await iroh.Iroh.init();
  final crypto = CryptoService();
  final (seed, public) = await crypto.generateEd25519KeyPair();
  final endpoint = await iroh.Endpoint.bindWithAddressLookup(
    resolve: (_) => null,
    relayMode: iroh.RelayMode.disabled,
  );
  MachineSession? session;
  IrohSessionLink? link;
  try {
    stdout.writeln(
      jsonEncode({
        'endpointId': endpoint.id.toHex(),
        'publicKey': base64Encode(public),
      }),
    );
    final line = await stdin
        .transform(utf8.decoder)
        .transform(const LineSplitter())
        .first;
    final config = jsonDecode(line) as Map<String, dynamic>;
    final expected = iroh.PublicKey.fromHex(config['endpointId'] as String);
    final connection = await endpoint.connect(
      iroh.EndpointAddr(
        expected,
        ipAddrs: (config['addresses'] as List).cast<String>(),
      ),
      utf8.encode('antgrid/peer/1'),
    );
    if (connection.remoteId != expected ||
        utf8.decode(connection.alpn) != 'antgrid/peer/1') {
      throw StateError('AUTHENTICATED_ENDPOINT_MISMATCH');
    }
    final stableId = connection.stableId;
    final (send, recv) = await connection.openBi();
    if (config['testCase'] == 'oversize-record' ||
        config['testCase'] == 'wrong-destination') {
      if (config['testCase'] == 'oversize-record') {
        await send.writeAll([255, 255, 255, 255]);
      } else {
        final raw = encodeRouteFrame(
          {'type': 'message', 'to': 'wrong-machine', 'channel': 'control'},
          Uint8List.fromList(utf8.encode('{}')),
          FrameKind.handshake,
        );
        final prefix = ByteData(4)..setUint32(0, raw.length, Endian.big);
        await send.writeAll([...prefix.buffer.asUint8List(), ...raw]);
      }
      await connection.closed();
      return;
    }
    final activeLink = link = IrohSessionLink(
      connection,
      send,
      recv,
      config['appId'] as String,
      config['machineId'] as String,
    );
    final active = session = MachineSession(
      relay: activeLink,
      machineDeviceId: config['machineId'] as String,
      handshaker: AppSessionHandshaker(
        relay: activeLink,
        crypto: crypto,
        machineDeviceId: config['machineId'] as String,
        phoneDeviceId: config['appId'] as String,
        agentEd25519PubB64: config['machinePublic'] as String,
        phoneEd25519Seed: seed,
        attemptTimeout: const Duration(seconds: 5),
        logger: (level, text, {fields}) => stderr.writeln('handshake: $text'),
      ),
    );
    var establishments = 0;
    final establishedSub = active.established.listen((_) {
      establishments++;
    });
    active.start();
    activeLink.start();
    Future<T> guarded<T>(Future<T> work) => Future.any([
      work,
      activeLink.failure.future.then<T>((_) => throw StateError('LINK_CLOSED')),
    ]);
    if ((config['testCase'] as String).startsWith('client-')) {
      active.ensureEstablished().ignore();
      try {
        await activeLink.failure.future;
        throw StateError('INVALID_PEER_ACCEPTED');
      } on StateError catch (error) {
        final expectedFailure = switch (config['testCase']) {
          'client-oversize-record' => 'INVALID_RECORD_LENGTH',
          'client-wrong-destination' => 'INVALID_ROUTE',
          'client-extra-stream' => 'EXTRA_STREAM',
          _ => throw StateError('UNKNOWN_CASE'),
        };
        if (error.message != expectedFailure) rethrow;
        stdout.writeln(
          jsonEncode({
            'check': 'peer-protocol-rejected',
            'testCase': config['testCase'],
          }),
        );
      }
      await establishedSub.cancel();
      return;
    }
    if (['bad-agent-key', 'bad-app-key'].contains(config['testCase'])) {
      try {
        await guarded(active.ensureEstablished());
        throw StateError('INVALID_IDENTITY_ESTABLISHED');
      } on HandshakeException {
        stdout.writeln(
          jsonEncode({
            'check': 'identity-rejected',
            'testCase': config['testCase'],
          }),
        );
      }
      await establishedSub.cancel();
      return;
    }
    await guarded(active.ensureEstablished());
    stderr.writeln('prototype: E2E established');
    if (config['testCase'] == 'extra-stream') {
      final (extra, _) = await connection.openBi();
      await extra.writeAll([1]);
      await connection.closed();
      await establishedSub.cancel();
      return;
    }
    if (config['testCase'] == 'disconnect') {
      try {
        await active.streamFor('0').request('qualification.drop');
        throw StateError('LOST_RPC_SUCCEEDED');
      } on RpcException catch (error) {
        if (error.code != 'E_SESSION_DOWN' || active.isEstablished) rethrow;
        stdout.writeln(
          jsonEncode({'check': 'pending-failed', 'code': error.code}),
        );
      }
      await establishedSub.cancel();
      return;
    }
    final streams = <String, String>{};
    for (final projectId in ['prototype-a', 'prototype-b']) {
      streams[projectId] = await guarded(
        active.bindProject(
          projectId,
          message('project:start', {'projectId': projectId}),
        ),
      );
      stderr.writeln('prototype: bound $projectId');
    }
    Future<void> terminal(String projectId, bool second) async {
      final transport = active.streamFor(streams[projectId]!);
      final done = Completer<void>();
      final expectedOutput = second ? '17993' : '27047';
      final sub = transport.messages.listen((event) {
        final frame = event.json;
        if (frame['type'] != 'terminal:frame') return;
        if (frame['terminalId'] != projectId) {
          if (!done.isCompleted)
            done.completeError(StateError('CROSS_PROJECT_FRAME'));
          return;
        }
        unawaited(
          transport.send(
            message('terminal:ack', {
              'terminalId': projectId,
              'runId': frame['runId'],
              'attachmentId': frame['attachmentId'],
              'sequence': frame['sequence'],
            }),
          ),
        );
        if ((frame['ansi'] as String).contains(expectedOutput) &&
            !done.isCompleted)
          done.complete();
      });
      try {
        await transport.send(
          message('terminal:subscribe', {
            'terminalId': projectId,
            'version': config['terminalVersion'],
            'requestId': uuid(),
          }),
        );
        final expression = second ? '947*19' : '731*37';
        final input = config['windows'] == true
            ? 'set /a $expression\r'
            : 'echo \$(($expression))\n';
        await transport.send(
          message('terminal:input', {'terminalId': projectId, 'data': input}),
        );
        await guarded(done.future.timeout(const Duration(seconds: 10)));
      } finally {
        await sub.cancel();
      }
    }

    await terminal('prototype-a', false);
    stderr.writeln('prototype: terminal a passed');
    await terminal('prototype-b', false);
    stderr.writeln('prototype: terminal b passed');
    final rebound = await active.bindProject(
      'prototype-a',
      message('project:start', {'projectId': 'prototype-a'}),
    );
    if (rebound != streams['prototype-a']) throw StateError('PROJECT_REDIALED');
    final rekeyed = active.established.first;
    for (var i = 0; i < 3; i++) {
      active.notifyRpcResult(timedOut: true);
    }
    await guarded(rekeyed.timeout(const Duration(seconds: 10)));
    stderr.writeln('prototype: rekeyed');
    await terminal('prototype-a', true);
    stdout.writeln(
      jsonEncode({
        'check': 'encrypted-terminal-pass',
        'establishments': establishments,
        'projects': streams.length,
        'sameConnection': stableId == connection.stableId,
      }),
    );
    await establishedSub.cancel();
  } finally {
    await session?.dispose();
    await link?.dispose();
    seed.fillRange(0, seed.length, 0);
    await endpoint.close();
    RustLib.dispose();
  }
}
